moon/controller.py

```diff
@@ -118,14 +124,38 @@ def send_speed_cmd(self, speed, angular_speed):
             self.virtual_bumper.update_desired_speed(speed, angular_speed)
         self.bus.publish('desired_speed', [round(speed*1000), round(math.degrees(angular_speed)*100)])
 
-    def send_request(self, cmd):
-        """Send ROS Service Request form a single place"""
-        self.publish('request', cmd)
-        while True:
-            dt, channel, data = self.listen()
-            if channel == 'response':
-                return data
-            print(dt, 'ignoring', channel)
+    def on_response(self, timestamp, data):
+        intoken, response = data
+        self.requests[intoken] = {
+            "response": response
+            }
+
+    def poll_responses(self, token, callback):
+        while self.time is None:
+            time.sleep(0.1)
+        request_sent = self.time
+        while self.time - request_sent < timedelta(seconds=5):
+            for i in self.requests:
+                if self.requests[token]["response"] is not None:
+                    response = self.requests[token]["response"]
+                    print(self.time, "controller:response received: token=%s, response=%s" % (token, response))
+                    self.requests.pop(token)
+                    if callback is not None:
+                        callback(response)
+                    return
+            time.sleep(0.1)
+        print(self.time, "controller: send_request timed out: token: %s" % token)
+
+        
+    def send_request(self, cmd, callback=None):
+        """Send ROS Service Request from a single place"""
+        token = secrets.token_hex(5)
```

Comment:
I am not familiar with `secrets` - does it always return the same code? I.e. is `replay` going to work?

Reply:
"secrets"-generated sequence of tokens is not reproducible. Do you want self.rand-based token to be used instead? It probably also doesn't need to be a (pseudo)random value, the token could as well be based on some incrementing counter.

Reply:
It was changed to `random` based string similar to other random events throughout the app.